File: 678-valid-parenthesis-string/valid-parenthesis-string.py
class Solution:
    def checkValidString(self, s: str) -> bool:
        left = 0
        right = 0
        
        for i in range(len(s)):
            if s[i] == '(' or s[i] == '*':
                left += 1
            else:
                left -= 1

            if s[len(s) - 1 - i] == ')' or s[len(s) - 1 - i] == '*':
                right += 1
            else:
                right -= 1

            if left < 0 or right < 0:
                return False
        
        return True

File: 678-valid-parenthesis-string/valid-parenthesis-string.cpp
class Solution {
public:
    bool checkValidString(string s) {        
        int left = 0, right = 0;
        for(int i = 0; i < s.length(); i++){
            if(s[i] == '(' || s[i] == '*') left++;
            else left--;

            if(s[s.length()-1-i] == ')' || s[s.length()-1-i] == '*') right++;
            else right--;

            if(left < 0 || right < 0) return false;
        }
        return true;
    }
};